core/templates/services/playthrough.service.ts

```diff
@@ -224,18 +235,25 @@ export class PlaythroughService {
       return;
     }
 
-    this.recordedLearnerActions.push(
-      this.learnerActionObjectFactory.createNewAnswerSubmitAction({
-        state_name: {value: stateName},
-        dest_state_name: {value: destStateName},
-        interaction_id: {value: interactionId},
-        submitted_answer: {value: answer},
-        feedback: {value: feedback},
-        time_spent_state_in_msecs: {value: 1000 * timeSpentInStateSecs}
-      }));
-
-    this.misTracker.recordStateTransition(destStateName);
-    this.cstTracker.recordStateTransition(destStateName);
+    if (this.recordedLearnerActions !== null) {
```

Comment:
Wouldn't it be an error if these variables were null at this point?

It might be worth analyzing the flow and possibly doing some light refactoring to make sure they aren't null at this point, rather than adding a guard. Ditto elsewhere.

More generally, putting null over the place just increases the complexity of the code and makes it less maintainable. If the types are telling you that something can be null when it really shouldn't be, that probably means there's a leak somewhere in the logic and it would be worth seeing whether that leak can be patched.

Reply:
These properties are explicitly intialize with null. Which indicates recording and scrutinizing playthroughs is not enabled/started.

Reply:
This doesn't address my question. I know that the previous author of the code initialized them with null. At that point, there wasn't a requirement to be careful about types, so people were a bit more loose about it. We shouldn't necessarily carry these patterns forward if we are increasing the level of quality checks on the code. The reasoning "it was already null" is therefore not sufficient.

Now that you are typing the code, I think these assumptions need to be reexamined and the code properly understood so that the right thing to do can be determined. The process of typing isn't just a mechanical process, it needs some proper analysis of what's going on.

Reply:
Absolutely correct, when i see these property are explicitly intialize with null, i blindly assume that their must be some reason. But now find that, their is no sense of using null.

Hence, removed.
Done